---
title: "Class 15: Transcriptomics and the analysis of RNA-Seq data"
author: "Sharon Tu"
date: "2/26/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Import countData and colData into R
```{r}
counts <- read.csv("data/airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("data/airway_metadata.csv", stringsAsFactors = FALSE)
```

```{r}
dim(counts)
head(counts)
```

```{r}
head(metadata)
```

```{r}
colnames(counts)
```

```{r}
metadata$id
```

Check for correspondence of metadata and countdata
```{r}
colnames(counts)[-1] == metadata$id
```

-Check for all of them at once by typing: 
```{r}
all(colnames(counts)[-1] == metadata$id)
```
> If all are true, it will read TRUE. If at least one is false, it will read FALSE.

First task is to find which columns in **counts** correspond to the control conditions (i.e. no drug!). Then we can calculate the mean count value for each gene.
```{r}
metadata$dex
```
```{r}
metadata$dex == "control"
control.inds <- metadata$dex == "control"
```

```{r}
head(counts[,control.inds])
```

```{r}
metadata$id[control.inds]
control.ids <- metadata$id[control.inds]
```

```{r}
head(counts[,control.ids])
```

This calculates the mean but not really a good way because if you decide you want to average over more or less than 4 numbers later, you would have to go back and change this code.
```{r}
rowSums(counts[,control.ids]) / 4
```

So, do this to calculate the mean counts value for all genes. You will be able to call this code for however many number of genes
```{r}
rowSums(counts[,control.ids]) / ncol(counts[control.ids])
control.mean <- rowSums(counts[,control.ids]) / ncol(counts[control.ids])
```

```{r}
names(control.mean) <- counts$ensgene
head(control.mean)
```

Now, do the same thing for the drug treated samples (i.e. *treated* columns)

```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/ ncol(counts[,treated$id]) 
names(treated.mean) <- counts$ensgene
```

Combine the mean count values for control and treated into a new data.frame 
```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

```{r}
colSums(meancounts)
```

Now we can plot the drug treated vs non-drug control...but where did all the dots go? 
```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log = "xy", xlab="log Control", ylab="log.Treated")
```

We can find candidate differentially expressed genes by looking for genes with a large change between control and dex-treated samples. We usually look at the log2 of the fold change, because this has better mathematical properties.

Here we calculate log2foldchange, add it to our meancounts data.frame and inspect the results either with the head() or the View() function for example.
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

```{r}
toy <- c(1,2,0,4,0)
toy == 0
which(toy == 0)
```

```{r}
toy <- meancounts[1:6,]
toy[,1:2]
which(toy[,1:2] == 0, arr.ind = TRUE)
unique(which(toy[,1:2] == 0, arr.ind=TRUE)[,1])
```

Remove the zero entry genes from meancounts to produce **mycounts**
```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

to.rm <- unique(zero.vals[,1])
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
sum(up.ind)
```

There are `sum(up.ind)` up regulated genes and `sum(down.ind)` down regulated genes
```{r}
head(mycounts[up.ind,])
```

## Adding annotation data
Use the **merge()** function to add annotation data from ensemble 
```{r}
anno <- read.csv("data/annotables_grch38.csv")
head(anno)
```

```{r}
mycounts.annotated <- merge(x=mycounts, y=anno, by.x="row.names", by.y="ensgene")
head(mycounts.annotated)
```

## DESeq2 analysis

Now we do this analysis properly with p-values!
```{r}
library(DESeq2)
```

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
head(res)
```

```{r}
summary(res)
```

```{r}
res01 <- results(dds, alpha=0.01)
summary(res01)
```

```{r}
#Setup  your point color vector 
mycols <- rep("gray", nrow(res01))
mycols[ abs(res01$log2FoldChange) > 2 ]  <- "red" 

inds <- (res01$padj < 0.01) & (abs(res01$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

#Volcano plot  with custom colors 
plot( res01$log2FoldChange,  -log(res01$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

https://github.com/bioboot/bimm143_W19_classwork/blob/master/class15/class15.md

