---
title: "class18"
author: "Sharon Tu"
date: "March 7, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Section 2: The GenomicDataCommons R package

```{r}
#source("https://bioconductor.org/biocLite.R")
#biocLite(c("GenomicDataCommons", "TCGAbiolinks", "maftools"))
```

```{r}
#library(GenomicDataCommons)
#library(TCGAbiolinks)
#library(maftools)
```

```{r}
GenomicDataCommons::status()
```

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

```{r}
x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec["TCGA-PAAD"] <- "red"

# Plot with 'log' for y axis and rotate labels with 'las'
#par(___)  
barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)
```

```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
```

```{r}
head(maf.file)
vars = read.maf(maf = maf.file, verbose = FALSE)
```

```{r}
plotmafSummary(vars)
```

```{r}
oncoplot(maf = vars, top = 10)
```

```{r}
# Oncoplot for our top 10 most frequently mutated genes
pdf("oncoplot_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
```

```{r}
oncostrip(maf=vars, genes=c("KRAS", "TP53"))
```

```{r}
lollipopPlot(vars, gene='KRAS')
```

```{r}
lollipopPlot(vars, gene='TP53')
```

## Hands on Part 2: Investigating cancer genomics datasets

```{r}
library(bio3d)
seqs <- read.fasta("~/Downloads/lecture18_sequences.fa")
seqs
```

```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```

```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names
```

```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```

```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")
```

** https://bioboot.github.io/bimm143_W19//class-material/lecture18_part2_example.Rmd ** 